docs/dockerfile_experimental.md

```diff
@@ -0,0 +1,128 @@
+As the experimental syntaxes may change in future revisions, you may want to pin the image to a specific revision.
+
+NOTE: `tonistiigi/dockerfile` is planned to be renamed to `docker/dockerfile-upstream` soon. See [#528](https://github.com/moby/buildkit/issues/528).
+
+## Experimental syntaxes
+
+### `RUN --mount=type=bind` (the default mount type)
+
+This mount type allows binding directories (read-only) in the context or in an image to the build container.
+
+|Option               |Description|
+|---------------------|-----------|
+|`target`(required)   | Mount path.|
+|`source`             | Source path in the `from`. Defaults to the root of the `from`.|
+|`from`               | Image name. Defaults to the build context.|
+
+
+### `RUN --mount=type=cache`
+
+This mount type allows the build container to cache directories for compilers and package managers.
+
+|Option               |Description|
+|---------------------|-----------|
+|`id`                 | ID of the cache.|
+|`target`(required)   | Mount path.|
+|`ro`,`readonly`      | Read-only if set.|
+|`sharing`            | Either one of `shared`, `private`, or `locked`. Defaults to `shared`. `shared` cache mount can be used concurrently by multiple writers. `private` creates a new mount if there are multiple writers. `locked` pauses second writer until first one releases the mount.|
+
+
+Example: cache Go packages
+
+```dockerfile
+# syntax = tonistiigi/dockerfile:master-experimental
+FROM golang
+...
+RUN --mount=type=cache,target=/root/.cache/go-build go build ...
+```
+
+Example: cache apt packages
+
+```dockerfile
+# syntax = tonistiigi/dockerfile:master-experimental
+FROM ubuntu
+RUN rm -f /etc/apt/apt.conf.d/docker-clean; echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
+RUN --mount=type=cache,target=/var/cache/apt --mount=type=cache,target=/var/lib/apt \
+  apt update && apt install -y gcc
+```
+
+### `RUN --mount=type=tmpfs`
+
+This mount type allows mounting tmpfs in the build container.
+
+|Option               |Description|
+|---------------------|-----------|
+|`target`(required)   | Mount path.|
+
+
+### `RUN --mount=type=secret`
+
+This mount type allows the build container to access credential files such as AWS keys without baking them into the image.
+
+|Option               |Description|
+|---------------------|-----------|
+|`id`                 | ID of the secret. Defaults to `path.Base(target)`.|
+|`target`             | Mount path. Defaults to `/run/secrets/ + path.Base(id)`.|
+|`required`           | If set to `true`, the instruction errors out when the secret is unavailable. Defaults to `false`.|
+
+
+Example: access to S3
+
+```dockerfile
+# syntax = tonistiigi/dockerfile:master-experimental
+FROM python:3
+RUN pip install awscli
+RUN --mount=type=secret,id=aws,target=/root/.aws/credentials aws s3 cp s3://... ...
+```
+
+```console
+$ buildctl build --frontend=dockerfile.v0 --local context=. --local dockerfile=. \
```

Comment:
Lets add the `docker build` example as well for now (not sure if `img` supports it).

Reply:
Can we defer until Docker v18.09 GA?

Reply:
We could just mention "(requires version v18.09+)"